[FIX] sale_coupon_advanced: remove discount lines with initial line

oerp-odoo opened the pull request:
Improving the way discount lines are handled, when related promotion order line is removed. It will not remove related discount line, if promotion line is removed (before, you would need to remove it manually).

oerp-odoo committed:
[FIX] sale_coupon_advanced: WIP

i-vyshnevska commented on sale_coupon_advanced/models/sale_order_line.py:
```diff
@@ -4,8 +4,37 @@
+        SaleCouponProgram = self.env[
+            'sale.coupon.program']
+        for line in self:
+            program = SaleCouponProgram.search([
```
```suggestion
            programs = SaleCouponProgram.search([
```
there can be a lot of them

oerp-odoo replied:
Right, but then I need to modify logic too. I somehow thought there would be only one. I need to be aware of that when filtering.

i-vyshnevska commented on sale_coupon_advanced/models/sale_order_line.py:
```diff
@@ -4,8 +4,37 @@
+        lines_to_remove = self.env['sale.order.line']
+        SaleCouponProgram = self.env[
+            'sale.coupon.program']
+        for line in self:
```
```suggestion
        for line in self.filtered(lambda self: self.forced_reward_line):
```

oerp-odoo replied:
As it is now, it should take care of all rewards, not just forced ones. Or we only want to remove related discount for forced lines?

StephaneMangin replied:
If the problem comes from our module, it should only be applied on the behavior it's made for. So @i-vyshnevska is right.

oerp-odoo replied:
Hm, its not that its the problem, it is functionality. I can only apply to to forced lines, but then it will gimp functionality. Is there a reason, user would want to keep non forced discount line, if related promotion line was removed? If not, then I see no reason to use it only for forced line, cause it could be useful in both cases.

i-vyshnevska commented on sale_coupon_advanced/models/sale_order_line.py:
```diff
@@ -4,8 +4,37 @@
+            )
+            order.no_code_promo_program_ids -= program
+            order.code_promo_program_id -= program
+        res = super(SaleOrderLine, self | lines_to_remove).unlink()
```
maybe try to call super method first and then remove just discount line

oerp-odoo replied:
Actually tried it, but getting same error.

i-vyshnevska reviewed (commented):
can you apply pre-commit, I want to check with which errors travis fails

StephaneMangin commented:
Can you indicate the card reference plz

oerp-odoo commented:
I was told to not indicate references on public PRs:)

StephaneMangin commented on sale_coupon_advanced/models/sale_order.py:
```diff
@@ -112,3 +112,47 @@ def _update_pricelist(self, pricelist):
         for line in self.order_line:
             line.product_id_change()
+
+    def write(self, vals):
```
I'm kind of perplex about the complexity of this behavior... Is it not a bit complicated to 'just' remove a discount line ?

StephaneMangin replied:
Does an unlink over the discount line not suffice ?

When i read that "But because we add it automatically (without needing to add that product, because we force it), it nullifies this standard behavior." : we would expect to 'fix' directly the behavior instead of adding a new level of code ? no ?

StephaneMangin replied:
Have you tried to remove the program associated to the discount line that remains after removing the product and apply the update procedure of the normal behavior ?

oerp-odoo replied:
Unfortunately its not that simple. First iteration implementation was done just like that. But there are side effects that happen when you try to unlink lines via Odoo interface (Form). It tries to use same line that is being deleted (by onchange spec). So we decided to make workaround that we clean up write if such things happens. You can check tests, specifically the one with Form. Thats special case, where all this complexity originates from.

oerp-odoo replied:
@StephaneMangin if by normal procedure you mean `recompute coupon lines`, then it wont work. As you already found my comment, it works the opposite of what we need. It adds missing lines, instead of removing. So in that case, we won't know if lines need to be added or removed.

ajaniszewska-dev commented:
> I was told to not indicate references on public PRs:)

then it is ok to write 2-3 sentences in PR description and explain what is the aim/scope of your PR, otherwise not always clear what you wanted to do.

StephaneMangin reviewed (approved):
Seen with @chroi @p-tombez and @yvaucher. We can't let this PR to block the release process. Can you just add a succinct explanation of the issue and add a FIXME to your write method?

oerp-odoo commented:
Sure, will do.

On Tue, Oct 13, 2020, 11:51 Stéphane Mangin <notifications@github.com>
wrote:

> *@StephaneMangin* approved this pull request.
>
> Seen with @chroi <https://github.com/chroi> @p-tombez
> <https://github.com/p-tombez> and @yvaucher <https://github.com/yvaucher>.
> We can't let this PR to block the release process. Can you just add a
> succinct explanation of the issue and add a FIXME to your write method?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/camptocamp/sale-workflow/pull/13#pullrequestreview-507209699>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB3TO6UXW5AQBAQGBKCTQETSKQIHRANCNFSM4SENR3LQ>
> .
>

oerp-odoo committed:
[FIX] sale_coupon_advanced: discount rem

When promotion order line is unlinked, related discount line is kept,
which is not the behavior user expects. It is expected to also remove
related discount line alongside promotion line.

Simple solution would be to override unlink to automatically include
related discount line when promotion line is unlinked. But it if unlink
happens via Odoo interface, it will mess up order lines, because Odoo
expects to keep discount line. This causes to conflicting write/unlink
where same order line is used in write and unlink.

Workaround is to make it work via write and unlink separately, where if
promotion line is detected to be unlinked via write, related discount
line is also handled to be deleted via write. If it is not detected via
write, then it is handled via unlink directly.

oerp-odoo commented:
@StephaneMangin still need one more approval. Or you already discussed this and I can proceed with merge?

chroi reviewed (approved):
lg
